syntax = "proto3";

package mlservice;

service MLService {
  rpc HealthCheck(HealthRequest) returns (HealthResponse);
  rpc ListModels(ListModelsRequest) returns (ListModelsResponse);
  rpc GetModelParams(GetModelParamsRequest) returns (GetModelParamsResponse);
  rpc TrainModel(TrainRequest) returns (TrainResponse);
  rpc ListTrainedModels(ListTrainedModelsRequest) returns (ListTrainedModelsResponse);
  rpc Predict(PredictRequest) returns (PredictResponse);
  rpc RetrainModel(RetrainRequest) returns (RetrainResponse);
  rpc DeleteModel(DeleteModelRequest) returns (DeleteModelResponse);
}

message HealthRequest {}

message HealthResponse {
  string status = 1;
  string timestamp = 2;
}

message ListModelsRequest {}

message ListModelsResponse {
  repeated string available_models = 1;
}

message GetModelParamsRequest {
  string model_name = 1;
}

message GetModelParamsResponse {
  string model_name = 1;
  map<string, string> default_hyperparameters = 2;
}

message TrainRequest {
  string model_name = 1;
  map<string, string> hyperparams = 2;
  string data_json = 3;
}

message TrainResponse {
  string model_id = 1;
  string model_name = 2;
  map<string, float> metrics = 3;
  string message = 4;
}

message ListTrainedModelsRequest {}

message ListTrainedModelsResponse {
  repeated TrainedModelInfo models = 1;
}

message TrainedModelInfo {
  string model_id = 1;
  string model_name = 2;
  string training_time = 3;
  map<string, float> metrics = 4;
}

message PredictRequest {
  string model_id = 1;
  string data_json = 2;
}

message PredictResponse {
  string model_id = 1;
  string model_name = 2;
  repeated float predictions = 3;
}

message RetrainRequest {
  string model_id = 1;
  map<string, string> hyperparams = 2;
  string data_json = 3;
}

message RetrainResponse {
  string model_id = 1;
  string model_name = 2;
  map<string, float> metrics = 3;
  string message = 4;
}

message DeleteModelRequest {
  string model_id = 1;
}

message DeleteModelResponse {
  string message = 1;
}